pkg/restore/crd.go

```diff
@@ -0,0 +1,220 @@
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func RestoreCrd(ctx context.Context, k8sclientset *k8sutil.Clientsets, operatorNamespace, clusterNamespace string, args []string) {
+	crd := args[0]
+	logging.Info("Detecting which resources to restore for crd %q", crd)
+	getCrName := `kubectl -n %s get %s -o jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.metadata.deletionGracePeriodSeconds}{"\n"}{end}' | awk '$2=="0" {print $1}'`
+	command := fmt.Sprintf(getCrName, clusterNamespace, crd)
+	crName := strings.TrimSpace(exec.ExecuteBashCommand(command))
+
+	if crName == "" {
+		logging.Info("Nothing to do here, no %q resources in deleted state", crd)
+		return
+	}
+
+	if len(args) == 2 {
+		crName = args[1]
+	}
+
+	logging.Info("Restoring CR %s", crName)
+
+	logging.Info("Scaling down the operator to 0")
+	err := k8sutil.SetDeploymentScale(ctx, k8sclientset.Kube, operatorNamespace, "rook-ceph-operator", 0)
+	if err != nil {
+		deploy, er := k8sutil.GetDeployment(ctx, k8sclientset.Kube, operatorNamespace, "rook-ceph-operator")
+		if er != nil && !apierrors.IsNotFound(er) {
+			logging.Fatal(er)
+		}
+		if deploy.Status.ReadyReplicas != 0 {
+			logging.Fatal(fmt.Errorf("Failed to scale down the operator deployment. %v", err))
+		}
+	}
+
+	logging.Info("Backing up kubernetes and crd resources")
+	crFileName := crd + "-" + crName + ".yaml"
+	getCrYamlContent := `kubectl -n %s get %s %s -oyaml > %s`
+	command = fmt.Sprintf(getCrYamlContent, clusterNamespace, crd, crName, crFileName)
+	exec.ExecuteBashCommand(command)
+	logging.Info("Backed up crd %s/%s in file %s", crd, crName, crFileName)
+
+	webhookConfigName := "rook-ceph-webhook"
+	logging.Info("Deleting validating webhook %s if present", webhookConfigName)
+	err = k8sclientset.Kube.AdmissionregistrationV1().ValidatingWebhookConfigurations().Delete(ctx, webhookConfigName, v1.DeleteOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		logging.Fatal(fmt.Errorf("failed to delete validating webhook %s. %v", webhookConfigName, err))
+	}
+
+	logging.Info("Fetching the UID for %s/%s", crd, crName)
+	getCrUID := `kubectl -n %s get %s %s -o 'jsonpath={.metadata.uid}'`
+	command = fmt.Sprintf(getCrUID, clusterNamespace, crd, crName)
+	uid := exec.ExecuteBashCommand(command)
+	logging.Info("Successfully fetched uid %s from %s/%s", uid, crd, crName)
+
+	removeOwnerRefOfUID(ctx, k8sclientset, operatorNamespace, clusterNamespace, uid)
+
+	logging.Info("Removing finalizers from %s/%s", crd, crName)
+	removeFinalizers := `kubectl -n %s patch %s/%s --type json --patch='[ { "op": "remove", "path": "/metadata/finalizers" } ]'`
+	command = fmt.Sprintf(removeFinalizers, clusterNamespace, crd, crName)
+	logging.Info(exec.ExecuteBashCommand(command))
+
+	logging.Info("Re-creating the CR %s from file %s created above", crd, crFileName)
+	recreateCR := `kubectl create -f %s`
+	command = fmt.Sprintf(recreateCR, crFileName)
+	logging.Info(exec.ExecuteBashCommand(command))
```

Comment:
Can this command fail? If so, is it fatal? We wouldn't want to continue and start the operator if it did fail.

Reply:
if the above commands, it will exit from their only. Method `ExecuteBashCommand` has check 
```
    stdout, err := cmd.Output()
	if err != nil {
		logging.Fatal(err)
	}
```

I think we are good here?